// codex-pets/src/installer.js
import { mkdir, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import path from "node:path";
import { defaultApiBase, petSlugPattern } from "./constants.js";
import { extractZipEntries } from "./zip.js";

const requiredPetFiles = ["pet.json", "spritesheet.webp"];

export async function addPet({
  petId,
  apiBase = process.env.CODEX_PETS_API_BASE ||
    process.env.PETSHARE_API_BASE ||
    defaultApiBase,
  codexDirectory = process.env.CODEX_HOME || path.join(homedir(), ".codex"),
  fetchImpl = fetch,
}) {
  const normalizedPetId = normalizePetId(petId);
  const baseUrl = normalizeApiBase(apiBase);
  const pet = await fetchPet(baseUrl, normalizedPetId, fetchImpl);
  return installPetPackage({
    baseUrl,
    pet,
    codexDirectory,
    fetchImpl,
  });
}

export async function addPetCollection({
  collectionSlug,
  apiBase = process.env.CODEX_PETS_API_BASE ||
    process.env.PETSHARE_API_BASE ||
    defaultApiBase,
  codexDirectory = process.env.CODEX_HOME || path.join(homedir(), ".codex"),
  fetchImpl = fetch,
  onPetInstalled,
} = {}) {
  const normalizedCollectionSlug = normalizeCollectionSlug(collectionSlug);
  const baseUrl = normalizeApiBase(apiBase);
  const collection = await fetchPetCollection(
    baseUrl,
    normalizedCollectionSlug,
    fetchImpl,
  );
  const installedPets = [];
  const totalPets = collection.pets.length;

  for (const [index, pet] of collection.pets.entries()) {
    const result = await installPetPackage({
      baseUrl,
      pet,
      codexDirectory,
      fetchImpl,
    });
    installedPets.push(result);
    onPetInstalled?.({
      pet: result,
      current: index + 1,
      total: totalPets,
    });
  }

  return {
    collection: collection.collection,
    pets: installedPets,
    total: installedPets.length,
  };
}

async function installPetPackage({ baseUrl, pet, codexDirectory, fetchImpl }) {
  const zipBytes = await fetchDownload(baseUrl, pet, fetchImpl);
  const entries = extractZipEntries(Buffer.from(zipBytes));
  const manifest = readManifest(entries, pet.id);
  const installPath = path.join(codexDirectory, "pets", manifest.id);

  await mkdir(installPath, { recursive: true });
  for (const fileName of requiredPetFiles) {
    const bytes = entries.get(fileName);
    if (!bytes) {
      throw new Error(`download did not include ${fileName}`);
    }
    await writeFile(path.join(installPath, fileName), bytes);
  }

  return {
    id: manifest.id,
    displayName: manifest.displayName,
    installPath,
  };
}

export function normalizePetId(petId) {
  const normalizedPetId = String(petId || "").trim();
  if (!petSlugPattern.test(normalizedPetId)) {
    throw new Error("pet id must be a lowercase slug like tiny-dino");
  }
  return normalizedPetId;
}

export function normalizeCollectionSlug(collectionSlug) {
  const normalizedCollectionSlug = String(collectionSlug || "").trim();
  if (!petSlugPattern.test(normalizedCollectionSlug)) {
    throw new Error("collection slug must be lowercase like cats");
  }
  return normalizedCollectionSlug;
}

export function normalizeApiBase(apiBase) {
  const trimmedApiBase = String(apiBase || "")
    .trim()
    .replace(/\/+$/, "");
  if (!trimmedApiBase) {
    throw new Error("API base URL is required");
  }

  const url = new URL(trimmedApiBase);
  if (!["http:", "https:"].includes(url.protocol)) {
    throw new Error("API base URL must use http or https");
  }
  return url.toString().replace(/\/+$/, "");
}

async function fetchPet(apiBase, petId, fetchImpl) {
  const response = await fetchImpl(`${apiBase}/api/pets/${petId}/share-data`);
  const body = await readJson(response);
  if (!response.ok) {
    throw new Error(
      body.error || `pet lookup failed with status ${response.status}`,
    );
  }
  if (!body.pet?.id) {
    throw new Error("pet lookup returned an invalid response");
  }
  return body.pet;
}

async function fetchPetCollection(apiBase, collectionSlug, fetchImpl) {
  const response = await fetchImpl(
    `${apiBase}/api/collections/${collectionSlug}`,
  );
  const body = await readJson(response);
  if (!response.ok) {
    throw new Error(
      body.error ||
        `pet collection lookup failed with status ${response.status}`,
    );
  }
  if (!body.collection?.slug || !Array.isArray(body.pets)) {
    throw new Error("pet collection lookup returned an invalid response");
  }

  return {
    collection: {
      ...body.collection,
      slug: normalizeCollectionSlug(body.collection.slug),
    },
    pets: body.pets.map(normalizePetSummary),
  };
}

async function fetchDownload(apiBase, pet, fetchImpl) {
  const downloadUrl = absoluteUrl(
    apiBase,
    pet.downloadUrl || `/api/pets/${pet.id}/download`,
  );
  const response = await fetchImpl(downloadUrl);
  if (!response.ok) {
    const body = await readJson(response);
    throw new Error(
      body.error || `pet download failed with status ${response.status}`,
    );
  }
  return response.arrayBuffer();
}

function absoluteUrl(apiBase, value) {
  if (value.startsWith("/")) {
    return `${apiBase}${value}`;
  }
  return new URL(value, `${apiBase}/`).toString();
}

async function readJson(response) {
  return response.json().catch(() => ({}));
}

function readManifest(entries, expectedPetId) {
  const manifestBytes = entries.get("pet.json");
  if (!manifestBytes) {
    throw new Error("download did not include pet.json");
  }

  const manifest = JSON.parse(new TextDecoder().decode(manifestBytes));
  if (manifest.id !== expectedPetId) {
    throw new Error("downloaded pet.json id did not match the requested pet");
  }
  if (manifest.spritesheetPath !== "spritesheet.webp") {
    throw new Error("downloaded pet.json has an invalid spritesheetPath");
  }
  return manifest;
}

function normalizePetSummary(pet) {
  if (!pet?.id) {
    throw new Error("pet collection lookup returned a pet without an id");
  }
  return {
    ...pet,
    id: normalizePetId(pet.id),
  };
}
